pkg/skaffold/build/parallel.go

```diff
@@ -21,96 +21,104 @@ import (
+
+	for i, a := range artifacts {
+		resChan := make(chan Result, 1)
+		resultChans[i] = resChan
 
-	// Run builds in //
-	for index := range artifacts {
-		i := index
+		// give each build a byte buffer to write its output to
 		lines := make(chan []byte, bufferedLinesPerArtifact)
 		outputs[i] = lines
 
-		r, w := io.Pipe()
-
-		// Log to the pipe, output will be collected and printed later
-		go func() {
-			// Make sure logs are printed in colors
-			var cw io.WriteCloser
-			if color.IsTerminal(out) {
-				cw = color.ColoredWriteCloser{WriteCloser: w}
-			} else {
-				cw = w
+		go func(artifact *latest.Artifact, c chan Result, lines chan []byte) {
+			res := &Result{
+				Target: *artifact,
 			}
-
-			color.Default.Fprintf(cw, "Building [%s]...\n", artifacts[i].ImageName)
-
-			event.BuildInProgress(artifacts[i].ImageName)
-
-			tag, present := tags[artifacts[i].ImageName]
-			if !present {
-				errs[i] = fmt.Errorf("unable to find tag for image %s", artifacts[i].ImageName)
-				event.BuildFailed(artifacts[i].ImageName, errs[i])
-			} else {
-				finalTags[i], errs[i] = buildArtifact(ctx, cw, artifacts[i], tag)
-				if errs[i] != nil {
-					event.BuildFailed(artifacts[i].ImageName, errs[i])
+			wg := &sync.WaitGroup{}
+
+			r, w := io.Pipe()
+
+			// Log to the pipe, output will be collected and printed later
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				// Make sure logs are printed in colors
+				var cw io.WriteCloser
+				if color.IsTerminal(out) {
+					cw = color.ColoredWriteCloser{WriteCloser: w}
+				} else {
+					cw = w
 				}
-			}
 
-			event.BuildComplete(artifacts[i].ImageName)
-			cw.Close()
-		}()
+				color.Default.Fprintf(cw, "Building [%s]...\n", artifact.ImageName)
+
+				event.BuildInProgress(artifact.ImageName)
+
+				tag, present := tags[artifact.ImageName]
+				if !present {
+					res.Error = fmt.Errorf("building [%s]: unable to find tag for image", artifact.ImageName)
+					event.BuildFailed(artifact.ImageName, res.Error)
+				} else {
+					bRes, err := buildArtifact(ctx, cw, artifact, tag)
+					if err != nil {
+						res.Error = err
+						event.BuildFailed(artifact.ImageName, err)
+					} else {
+						res.Result = Artifact{
+							ImageName: artifact.ImageName,
+							Tag:       bRes,
+						}
+					}
+				}
+				event.BuildComplete(artifact.ImageName)
+				cw.Close()
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				scanner := bufio.NewScanner(r)
+				for scanner.Scan() {
+					lines <- scanner.Bytes()
+				}
+				close(lines)
+			}()
 
-		go func() {
-			scanner := bufio.NewScanner(r)
-			for scanner.Scan() {
-				lines <- scanner.Bytes()
-			}
-			close(lines)
-		}()
+			wg.Wait() // wait for build to finish and output to be processed
+			c <- *res // send the result back through the callback channel
+		}(a, resChan, lines)
 	}
 
-	// Print logs and collect results in order.
-	var built []Artifact
-
-	for i, artifact := range artifacts {
+	for i := range artifacts {
```

Comment:
Because we left this here, the parallel builder won't return until all the outputs are consumed - which means that the eventing logic is going to be "batched" until all builds are done. 

So: 

1. we should write a test to ensure "realtimeness" of the events somehow 
2. we should move the output handling to another go routine